AI/BattleAI/AttackPossibility.cpp

```diff
@@ -44,24 +44,30 @@ void DamageCache::buildObstacleDamageCache(std::shared_ptr<HypotheticBattle> hb,
 		if(!triggerIsNegative)
 			continue;
 
-		const auto * hero = hb->battleGetFightingHero(spellObstacle->casterSide);
-		auto caster = spells::ObstacleCasterProxy(hb->getSidePlayer(spellObstacle->casterSide), hero, *spellObstacle);
+		std::unique_ptr<spells::BattleCast> cast = nullptr;
+		if(spellObstacle->obstacleType == SpellCreatedObstacle::EObstacleType::SPELL_CREATED)
```

Comment:
Probably we can skip whole iteration in case we have no cast there

Reply:
Maybe we can just replace battleGetFightingHero with getSideHero.

Reply:
`getSideHero` uses it anyway:
```
const CGHeroInstance * BattleProxy::getSideHero(BattleSide side) const
{
	return subject->battleGetFightingHero(side);
}
```

Reply:
> Probably we can skip whole iteration in case we have no cast there. Just continue. But this should act on moats as well. I do not know if SPELL_CREATED will work for moats.

Moats have separate type:
```
	enum EObstacleType : ui8
	{
		//ABSOLUTE needs an underscore because it's a Win
		USUAL, ABSOLUTE_OBSTACLE, SPELL_CREATED, MOAT
	};
```